File: internal/checker/checker.go
package checker

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/anthropics/claude-3-5-sonnet-20241022-experiments/internal/environment"
	"github.com/anthropics/claude-3-5-sonnet-20241022-experiments/internal/task"
)

// CheckResult represents the result of checking a criterion
type CheckResult struct {
	Passed bool
	Detail string
}

// Checker evaluates task criteria against environment context
type Checker struct {
	env *environment.EnvironmentContext
}

// New creates a new Checker with the given environment context
func New(env *environment.EnvironmentContext) *Checker {
	return &Checker{env: env}
}

// CheckCriterion evaluates a single criterion and returns the result
func (c *Checker) CheckCriterion(criterion task.Criterion) CheckResult {
	switch criterion.Type {
	case "file_exists":
		return c.checkFileExists(criterion.FilePaths)
	case "command_exits_zero":
		return c.checkCommandExitsZero(criterion.Command)
	case "output_contains":
		return c.checkOutputContains(criterion.Command, criterion.OutputContains)
	default:
		return CheckResult{
			Passed: false,
			Detail: fmt.Sprintf("unknown criterion type: %s", criterion.Type),
		}
	}
}

// CheckAllCriteria evaluates all criteria independently and returns results
func (c *Checker) CheckAllCriteria(criteria []task.Criterion) []CheckResult {
	results := make([]CheckResult, len(criteria))
	for i, criterion := range criteria {
		results[i] = c.CheckCriterion(criterion)
	}
	return results
}

func (c *Checker) checkFileExists(paths []string) CheckResult {
	var missingPaths []string
	
	for _, path := range paths {
		fullPath := filepath.Join(c.env.WorkspaceDir, path)
		if _, err := os.Stat(fullPath); os.IsNotExist(err) {
			missingPaths = append(missingPaths, path)
		}
	}
	
	if len(missingPaths) == 0 {
		return CheckResult{Passed: true}
	}
	
	return CheckResult{
		Passed: false,
		Detail: fmt.Sprintf("missing paths: %s", strings.Join(missingPaths, ", ")),
	}
}

func (c *Checker) checkCommandExitsZero(command string) CheckResult {
	cmd := exec.Command("sh", "-c", command)
	cmd.Dir = c.env.WorkspaceDir
	cmd.Env = append(os.Environ(), fmt.Sprintf("HOME=%s", c.env.HomeDir))
	
	var stderr strings.Builder
	cmd.Stderr = &stderr
	
	err := cmd.Run()
	if err != nil {
		return CheckResult{
			Passed: false,
			Detail: stderr.String(),
		}
	}
	
	return CheckResult{Passed: true}
}

func (c *Checker) checkOutputContains(command string, contains []string) CheckResult {
	cmd := exec.Command("sh", "-c", command)
	cmd.Dir = c.env.WorkspaceDir
	cmd.Env = append(os.Environ(), fmt.Sprintf("HOME=%s", c.env.HomeDir))
	
	var stdout strings.Builder
	cmd.Stdout = &stdout
	
	err := cmd.Run()
	if err != nil {
		return CheckResult{
			Passed: false,
			Detail: fmt.Sprintf("command failed: %v", err),
		}
	}
	
	output := stdout.String()
	var missingStrings []string
	
	for _, str := range contains {
		if !strings.Contains(output, str) {
			missingStrings = append(missingStrings, str)
		}
	}
	
	if len(missingStrings) == 0 {
		return CheckResult{Passed: true}
	}
	
	return CheckResult{
		Passed: false,
		Detail: fmt.Sprintf("missing strings in output: %s", strings.Join(missingStrings, ", ")),
	}
}

File: src/token_miser/checker.py
"""Success criteria verification."""

from __future__ import annotations

import os
import subprocess
from dataclasses import dataclass

from token_miser.environment import EnvironmentContext
from token_miser.task import Criterion


@dataclass
class CheckResult:
    passed: bool
    detail: str = ""


def check_criterion(criterion: Criterion, env: EnvironmentContext) -> CheckResult:
    """Evaluate a single success criterion."""
    match criterion.type:
        case "file_exists":
            return _check_file_exists(criterion, env)
        case "command_exits_zero" | "command_succeeds":
            return _check_command_exits_zero(criterion, env)
        case "output_contains":
            return _check_output_contains(criterion, env)
        case _:
            return CheckResult(passed=False, detail=f"Unknown criterion type: {criterion.type}")


def check_all_criteria(criteria: list[Criterion], env: EnvironmentContext) -> list[CheckResult]:
    """Evaluate all criteria independently (no short-circuit)."""
    return [check_criterion(c, env) for c in criteria]


def _check_file_exists(criterion: Criterion, env: EnvironmentContext) -> CheckResult:
    missing = [p for p in criterion.paths if not os.path.exists(os.path.join(env.workspace_dir, p))]
    if missing:
        return CheckResult(passed=False, detail=f"Missing files: {', '.join(missing)}")
    return CheckResult(passed=True)


def _check_command_exits_zero(criterion: Criterion, env: EnvironmentContext) -> CheckResult:
    run_env = os.environ.copy()
    run_env["HOME"] = env.home_dir
    try:
        proc = subprocess.run(
            ["sh", "-c", criterion.command],
            cwd=env.workspace_dir,
            env=run_env,
            capture_output=True,
            text=True,
            timeout=120,
        )
        if proc.returncode == 0:
            return CheckResult(passed=True)
        stderr = proc.stderr.strip() if proc.stderr else "(no output)"
        return CheckResult(passed=False, detail=f"Exit code {proc.returncode}: {stderr}")
    except subprocess.TimeoutExpired:
        return CheckResult(passed=False, detail="Command timed out after 120s")


def _check_output_contains(criterion: Criterion, env: EnvironmentContext) -> CheckResult:
    run_env = os.environ.copy()
    run_env["HOME"] = env.home_dir
    try:
        proc = subprocess.run(
            ["sh", "-c", criterion.command],
            cwd=env.workspace_dir,
            env=run_env,
            capture_output=True,
            text=True,
            timeout=120,
        )
        missing = [s for s in criterion.contains if s not in proc.stdout]
        if missing:
            return CheckResult(passed=False, detail=f"Missing in output: {', '.join(missing)}")
        if proc.returncode != 0:
            detail = f"Command exited {proc.returncode} (output matched but command failed)"
            return CheckResult(passed=False, detail=detail)
        return CheckResult(passed=True)
    except subprocess.TimeoutExpired:
        return CheckResult(passed=False, detail="Command timed out after 120s")
